bgpd/bgp_ecommunity.c

```diff
@@ -1026,3 +1026,144 @@ int ecommunity_fill_pbr_action(struct ecommunity_val *ecom_eval,
 		return -1;
 	return 0;
 }
+
+static unsigned int bgp_aggr_ecommunity_hash_key(void *data)
+{
+	struct ecommunity *ecommunity = data;
+
+	return jhash(ecommunity->val, ecommunity->size, 0xdcba4321);
+}
+
+static bool bgp_aggr_ecommunity_hash_cmp(const void *n1, const void *n2)
+{
+	const struct ecommunity *ecom1 = n1;
+	const struct ecommunity *ecom2 = n2;
+
+	if (ecom1 == NULL && ecom2 == NULL)
+		return true;
+
+	if (ecom1 == NULL || ecom2 == NULL)
+		return false;
+
+	return (ecom1->size == ecom2->size
+		&& memcmp(ecom1->val, ecom2->val, ecom1->size * ECOMMUNITY_SIZE)
+			   == 0);
+}
+
+static struct ecommunity *bgp_aggr_ecommunity_lookup(
+						struct bgp_aggregate *aggregate,
+						struct ecommunity *ecommunity)
+{
+	return hash_lookup(aggregate->ecommunity_hash, ecommunity);
+}
+
+static void *bgp_aggr_ecommunty_hash_alloc(void *p)
+{
+	struct ecommunity *ref = (struct ecommunity *)p;
+	struct ecommunity *ecommunity = NULL;
+
+	ecommunity = ecommunity_dup(ref);
+	return ecommunity;
+}
+
+static void bgp_aggr_ecommunity_prepare(struct hash_backet *hb, void *arg)
+{
+	struct ecommunity *ecommerge = NULL;
+	struct ecommunity *hb_ecommunity = hb->data;
+	struct ecommunity **aggr_ecommunity = arg;
+
+	if (*aggr_ecommunity) {
+		ecommerge = ecommunity_merge(*aggr_ecommunity, hb_ecommunity);
+		*aggr_ecommunity = ecommunity_uniq_sort(ecommerge);
+		ecommunity_free(&ecommerge);
+	} else
+		*aggr_ecommunity = ecommunity_dup(hb_ecommunity);
+}
+
+void bgp_aggr_ecommunity_remove(void *arg)
+{
+	struct ecommunity *ecommunity = arg;
+
+	ecommunity_free(&ecommunity);
+}
+
+void bgp_compute_aggregate_ecommunity(struct bgp_aggregate *aggregate,
+				      struct ecommunity *ecommunity)
+{
+	struct ecommunity *aggr_ecommunity = NULL;
+
+	if ((aggregate == NULL) || (ecommunity == NULL))
+		return;
+
+	/* Create hash if not already created.
+	 */
+	if (aggregate->ecommunity_hash == NULL) {
+		aggregate->ecommunity_hash = hash_create(
+					bgp_aggr_ecommunity_hash_key,
+					bgp_aggr_ecommunity_hash_cmp,
+					"BGP Aggregator ecommunity hash");
+		assert(aggregate->ecommunity_hash != NULL);
```

Comment:
This is impossible, what is the assert for?

Reply:
I've removed this check.